File: sdk/go/cytindexer/catalog_io.go
package cytindexer

import (
	"os"
	"path/filepath"
	"strings"
)

// ResolveWriteCatalogPaths resolves output directory and prune flag from optional overrides.
func ResolveWriteCatalogPaths(outputDir string, prune *bool) (string, bool) {
	cfg := PathSnapshot()
	dir := outputDir
	if dir == "" {
		dir = cfg.DefaultCatalogDir
	}
	doPrune := cfg.WriteCatalogPrune
	if prune != nil {
		doPrune = *prune
	}
	return dir, doPrune
}

// WriteCatalogIndexResolved writes a catalog index with optional dir/prune overrides.
func WriteCatalogIndexResolved(index *CatalogIndex, outputDir string, prune *bool) error {
	dir, doPrune := ResolveWriteCatalogPaths(outputDir, prune)
	return WriteCatalogIndex(index, dir, doPrune)
}

// WriteCatalogIndex writes catalog files to disk.
func WriteCatalogIndex(index *CatalogIndex, outputDir string, prune bool) error {
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Join(outputDir, "schemas"), 0o755); err != nil {
		return err
	}

	outputMap := make(map[string]string)
	for relPath, content := range index.Files {
		outputMap[filepath.Join(outputDir, relPath)] = content
	}

	if err := applyOutputs(outputMap); err != nil {
		return err
	}
	if prune {
		expected := make(map[string]struct{}, len(outputMap))
		for p := range outputMap {
			expected[p] = struct{}{}
		}
		return pruneStaleFiles(outputDir, expected)
	}
	return nil
}

func applyOutputs(outputMap map[string]string) error {
	for path, content := range outputMap {
		if existing, err := os.ReadFile(path); err == nil && string(existing) == content {
			continue
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return err
		}
		if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
			return err
		}
	}
	return nil
}

func shouldSkipHidden(path, root string) bool {
	rel, err := filepath.Rel(root, path)
	if err != nil {
		return false
	}
	for _, part := range strings.Split(filepath.ToSlash(rel), "/") {
		if strings.HasPrefix(part, ".") {
			return true
		}
	}
	return false
}

func pruneStaleFiles(root string, expected map[string]struct{}) error {
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return nil
	}

	var allPaths []string
	if err := collectPaths(root, root, &allPaths); err != nil {
		return err
	}

	for _, path := range allPaths {
		if shouldSkipHidden(path, root) {
			continue
		}
		if info, err := os.Stat(path); err == nil && !info.IsDir() {
			if _, ok := expected[path]; !ok {
				_ = os.Remove(path)
			}
		}
	}

	// Remove empty directories deepest-first.
	sortPathsByDepthDesc(allPaths)
	for _, path := range allPaths {
		if shouldSkipHidden(path, root) {
			continue
		}
		if info, err := os.Stat(path); err == nil && info.IsDir() {
			entries, _ := os.ReadDir(path)
			if len(entries) == 0 {
				_ = os.Remove(path)
			}
		}
	}
	return nil
}

func sortPathsByDepthDesc(paths []string) {
	for i := 0; i < len(paths); i++ {
		for j := i + 1; j < len(paths); j++ {
			if strings.Count(paths[j], string(os.PathSeparator)) > strings.Count(paths[i], string(os.PathSeparator)) {
				paths[i], paths[j] = paths[j], paths[i]
			}
		}
	}
}

func collectPaths(dir, root string, out *[]string) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		path := filepath.Join(dir, entry.Name())
		*out = append(*out, path)
		if entry.IsDir() {
			if err := collectPaths(path, root, out); err != nil {
				return err
			}
		}
	}
	return nil
}

File: sdk/rust/cyt-indexer/src/catalog_io.rs
use crate::build::CatalogIndex;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

/// Resolve output directory and prune flag from optional overrides and path config.
pub fn resolve_write_catalog_paths(
    output_dir: Option<&Path>,
    prune: Option<bool>,
) -> (PathBuf, bool) {
    let cfg = crate::paths::snapshot();
    let dir = output_dir
        .map(Path::to_path_buf)
        .unwrap_or(cfg.default_catalog_dir);
    let prune = prune.unwrap_or(cfg.write_catalog_prune);
    (dir, prune)
}

/// Write catalog index using optional output dir / prune (defaults from [`crate::paths`]).
///
/// # Errors
/// Returns an error string when the output directory cannot be created or catalog files cannot be written.
pub fn write_catalog_index_resolved(
    index: &CatalogIndex,
    output_dir: Option<&Path>,
    prune: Option<bool>,
) -> Result<(), String> {
    let (dir, prune) = resolve_write_catalog_paths(output_dir, prune);
    write_catalog_index(index, &dir, prune)
}

/// Write catalog index files under `output_dir`, optionally pruning stale entries.
///
/// # Errors
/// Returns an error string when the output directory cannot be created, files cannot be written, or pruning fails.
pub fn write_catalog_index(
    index: &CatalogIndex,
    output_dir: &Path,
    prune: bool,
) -> Result<(), String> {
    let root = output_dir;
    fs::create_dir_all(root).map_err(|e| e.to_string())?;
    let schemas = root.join("schemas");
    fs::create_dir_all(&schemas).map_err(|e| e.to_string())?;

    let mut output_map: HashMap<PathBuf, String> = HashMap::new();
    for (rel_path, content) in &index.files {
        output_map.insert(root.join(rel_path), content.clone());
    }

    apply_outputs(&output_map)?;
    if prune {
        let expected: HashSet<PathBuf> = output_map.keys().cloned().collect();
        prune_stale_files(root, &expected)?;
    }
    Ok(())
}

fn apply_outputs(output_map: &HashMap<PathBuf, String>) -> Result<(), String> {
    for (path, content) in output_map {
        if path.exists()
            && let Ok(existing) = fs::read_to_string(path)
            && existing == *content
        {
            continue;
        }
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).map_err(|e| e.to_string())?;
        }
        fs::write(path, content).map_err(|e| e.to_string())?;
    }
    Ok(())
}

fn should_skip_hidden(path: &Path, root: &Path) -> bool {
    path.strip_prefix(root).ok().is_some_and(|rel| {
        rel.components()
            .any(|c| c.as_os_str().to_string_lossy().starts_with('.'))
    })
}

fn prune_stale_files(root: &Path, expected_paths: &HashSet<PathBuf>) -> Result<(), String> {
    if !root.exists() {
        return Ok(());
    }

    let mut all_paths: Vec<PathBuf> = Vec::new();
    collect_paths(root, &mut all_paths)?;

    for path in &all_paths {
        if should_skip_hidden(path, root) {
            continue;
        }
        if path.is_file() && !expected_paths.contains(path) {
            fs::remove_file(path).map_err(|e| e.to_string())?;
        }
    }

    all_paths.sort_by_key(|p| std::cmp::Reverse(p.components().count()));
    for path in all_paths {
        if should_skip_hidden(&path, root) {
            continue;
        }
        if path.is_dir() && path.read_dir().is_ok_and(|mut d| d.next().is_none()) {
            let _ = fs::remove_dir(path);
        }
    }
    Ok(())
}

fn collect_paths(dir: &Path, out: &mut Vec<PathBuf>) -> Result<(), String> {
    if !dir.is_dir() {
        return Ok(());
    }
    for entry in fs::read_dir(dir).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let path = entry.path();
        out.push(path.clone());
        if path.is_dir() {
            collect_paths(&path, out)?;
        }
    }
    Ok(())
}
